File: bindings/typescript/test/config.test.ts
import { describe, it, expect } from 'vitest'
import { createRomcal, RomcalError } from '../src/index.js'

describe('Romcal configuration', () => {
  it('should use default configuration', async () => {
    const romcal = await createRomcal()

    expect(romcal.calendar).toBe('general_roman')
    expect(romcal.locale).toBe('en')
    expect(romcal.epiphanyOnSunday).toBe(false)
    expect(romcal.corpusChristiOnSunday).toBe(true)
    expect(romcal.ascensionOnSunday).toBe(false)
    expect(romcal.easterCalculationType).toBe('gregorian')
    expect(romcal.context).toBe('gregorian')
  })

  it('should accept calendar and locale as arguments', async () => {
    const romcal = await createRomcal('france', 'fr')

    expect(romcal.calendar).toBe('france')
    expect(romcal.locale).toBe('fr')
  })

  it('should accept partial configuration object', async () => {
    const romcal = await createRomcal({
      calendar: 'united_states',
      locale: 'en',
      epiphanyOnSunday: true,
      ascensionOnSunday: true,
    })

    expect(romcal.calendar).toBe('united_states')
    expect(romcal.locale).toBe('en')
    expect(romcal.epiphanyOnSunday).toBe(true)
    expect(romcal.ascensionOnSunday).toBe(true)
    // Default values preserved
    expect(romcal.corpusChristiOnSunday).toBe(true)
  })

  it('should throw RomcalError for invalid easter calculation type', async () => {
    await expect(createRomcal({ easterCalculationType: 'invalid' as never })).rejects.toThrow(
      RomcalError,
    )
    await expect(createRomcal({ easterCalculationType: 'invalid' as never })).rejects.toThrow(
      "Invalid easter_calculation_type: 'invalid'",
    )
  })

  it('should throw RomcalError for invalid context', async () => {
    await expect(createRomcal({ context: 'invalid' as never })).rejects.toThrow(RomcalError)
    await expect(createRomcal({ context: 'invalid' as never })).rejects.toThrow(
      "Invalid context: 'invalid'",
    )
  })
})

File: bindings/kotlin/src/test/kotlin/com/romcal/ConfigTest.kt
package com.romcal

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ConfigTest {

    @Test
    fun `default configuration should have expected values`() {
        Romcal().use { romcal ->
            assertEquals("general_roman", romcal.calendar)
            assertEquals("en", romcal.locale)
            assertFalse(romcal.epiphanyOnSunday)
            assertFalse(romcal.ascensionOnSunday)
            assertTrue(romcal.corpusChristiOnSunday)
            assertEquals(EasterCalculationType.GREGORIAN, romcal.easterCalculationType)
            assertEquals(CalendarContext.GREGORIAN, romcal.context)
        }
    }

    @Test
    fun `should accept calendar and locale arguments`() {
        Romcal(calendar = "france", locale = "fr").use { romcal ->
            assertEquals("france", romcal.calendar)
            assertEquals("fr", romcal.locale)
        }
    }

    @Test
    fun `should accept partial configuration with defaults`() {
        Romcal(
            epiphanyOnSunday = true,
            ascensionOnSunday = true,
            corpusChristiOnSunday = false
        ).use { romcal ->
            assertTrue(romcal.epiphanyOnSunday)
            assertTrue(romcal.ascensionOnSunday)
            assertFalse(romcal.corpusChristiOnSunday)
            // Default values should still apply
            assertEquals("general_roman", romcal.calendar)
            assertEquals("en", romcal.locale)
        }
    }

    @Test
    fun `should accept easter calculation type configuration`() {
        Romcal(easterCalculationType = EasterCalculationType.JULIAN).use { romcal ->
            assertEquals(EasterCalculationType.JULIAN, romcal.easterCalculationType)
        }
    }

    @Test
    fun `should accept context configuration`() {
        Romcal(context = CalendarContext.LITURGICAL).use { romcal ->
            assertEquals(CalendarContext.LITURGICAL, romcal.context)
        }
    }
}
